src/v/raft/mutex_buffer.h

```diff
@@ -0,0 +1,130 @@
+#pragma once
+#include "raft/types.h"
+#include "ssx/future-util.h"
+#include "utils/mutex.h"
+
+#include <seastar/core/condition-variable.hh>
+#include <seastar/core/do_with.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/gate.hh>
+#include <seastar/core/loop.hh>
+#include <seastar/core/semaphore.hh>
+#include <seastar/core/shared_ptr.hh>
+
+namespace raft::details {
+
+template<typename Request, typename Response>
+class mutex_buffer {
+public:
+    explicit mutex_buffer(mutex& m, size_t max_buffered_elements)
+      : _mutex(m)
+      , _max_buffered(max_buffered_elements) {}
+
+    ss::future<Response> enqueue(Request&& r) {
+        if (_gate.is_closed()) {
+            return ss::make_exception_future<Response>(
+              ss::gate_closed_exception{});
+        }
+        // we do not use a semaphore as we do not want to wait, waiting may
+        // cause requests reordering
+        if (_requests.size() >= _max_buffered) {
+            return ss::make_exception_future<Response>(
+              std::overflow_error("max buffered entries count reached."));
+        }
+        ss::promise<Response> p;
+        auto f = p.get_future();
+        _requests.push_back(std::move(r));
+        _responsens.push_back(std::move(p));
+        _enequeued.signal();
+        return f;
+    }
+    // clang-format off
+    template<typename Func>
+    CONCEPT(requires requires(Func f, Request req) {
+        { f(std::move(req)) } -> ss::future<Response>;
+    })
+    // clang-format on
+    void start(Func&& f);
```

Comment:
Can you add comments on how to use this class. It is not clear from the code. For example start() doesn't handle gate closed exception.

Reply:
I will add a comment